src/pkg/services/m365/api/teams.go

```diff
@@ -0,0 +1,161 @@
+package api
+
+import (
+	"context"
+
+	"github.com/alcionai/clues"
+	"github.com/alcionai/corso/src/internal/common/str"
+	"github.com/alcionai/corso/src/internal/common/tform"
+	"github.com/alcionai/corso/src/internal/m365/graph"
+	"github.com/alcionai/corso/src/pkg/fault"
+	msgraphgocore "github.com/microsoftgraph/msgraph-sdk-go-core"
+	"github.com/microsoftgraph/msgraph-sdk-go/models"
+)
+
+const teamService = "Team"
+
+// ---------------------------------------------------------------------------
+// controller
+// ---------------------------------------------------------------------------
+
+func (c Client) Groups() Groups {
+	return Groups{c}
+}
+
+// On creation of each Microsoft Teams a corrsponding group gets created from them.
+// Most of the information like events, drive and mail info will be fetched directly
+// from groups. So we pull in group and process only the once which are associated with
+// a team for further proccessing of teams.
+
+// Teams is an interface-compliant provider of the client.
+type Groups struct {
+	Client
+}
+
+// GetAll retrieves all groups.
+func (c Groups) GetAll(
+	ctx context.Context,
+	filterTeams bool,
+	errs *fault.Bus,
+) ([]models.Groupable, error) {
+	service, err := c.Service()
+	if err != nil {
+		return nil, err
+	}
+
+	return getGroups(ctx, filterTeams, errs, service)
+}
+
+// GetAll retrieves all groups.
+func getGroups(
+	ctx context.Context,
+	filterTeams bool,
+	errs *fault.Bus,
+	service graph.Servicer,
+) ([]models.Groupable, error) {
+
+	resp, err := service.Client().Groups().Get(ctx, nil)
+	if err != nil {
+		return nil, graph.Wrap(ctx, err, "getting all groups")
+	}
+
+	iter, err := msgraphgocore.NewPageIterator[models.Groupable](
+		resp,
+		service.Adapter(),
+		models.CreateTeamCollectionResponseFromDiscriminatorValue)
+	if err != nil {
+		return nil, graph.Wrap(ctx, err, "creating groups iterator")
+	}
+
+	var (
+		groups = make([]models.Groupable, 0)
+		el     = errs.Local()
+	)
+
+	iterator := func(item models.Groupable) bool {
+		if el.Failure() != nil {
+			return false
+		}
+
+		err := ValidateGroup(item)
+		if err != nil {
+			el.AddRecoverable(ctx, graph.Wrap(ctx, err, "validating groups"))
+		} else {
+			isTeam := IsTeam(item)
+			if !filterTeams || isTeam {
+				groups = append(groups, item)
+			}
+		}
+
+		return true
+	}
+
+	if err := iter.Iterate(ctx, iterator); err != nil {
+		return nil, graph.Wrap(ctx, err, "iterating all groups")
+	}
+
+	return groups, el.Failure()
+}
+
+func IsTeam(g models.Groupable) bool {
+	if g.GetAdditionalData()["resourceProvisioningOptions"] != nil {
+		val, _ := tform.AnyValueToT[[]any]("resourceProvisioningOptions", g.GetAdditionalData())
+		for _, v := range val {
+			s, err := str.AnyToString(v)
+			if err != nil {
+				return false
+			}
+			if s == teamService {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// GetID retrieves team by groupID/teamID.
+func (c Groups) GetByID(
+	ctx context.Context,
+	identifier string,
```

Comment:
Can this fetch a team using a team ID, or does it need a group ID that is different from the team id?

Reply:
With my understanding/investigation - value of groupID created for the team is same as TeamID.

Reply:
Sounds good.  In that case, we'll want this func to accept any variation of identifier we use for the team.  For example: sites can provide either the site id or the webURL, users can provide either user id or principal name.  Given that not all groups are teams, we might want to make a `GetTeamByID` func instead.

Though, now that I think about it, maybe betwen this and the GetAllTeams suggestion, this is a good time to swap this client from `Groups` to `Teams`, and have the teams api call groups under the hood?  It'll probably be clearer for us to make calls like `ac.Teams().GetByID()` and `ac.Teams().GetGroupByID()` than it would be to have lots of `ac.Groups().GetFooByID()` variations.

Reply:
Done.